ios/engine/KMEI/KeymanEngine/Classes/Manager.swift

```diff
@@ -0,0 +1,3018 @@
+                        font: DefaultKeyboard.keyboardFont, oskFont: nil)
+      }
+      return false
+    }
+
+    self.languageID = languageID
+    self.keyboardID = keyboardID
+    let key = "\(languageID)_\(keyboardID)"
+
+    let userKeyboards = activeUserDefaults().array(forKey: Key.userKeyboardsList)
+    let index = indexForUserKeyboard(withID: keyboardID, languageID: languageID)
+
+    let kbDict: [String: String]?
+    if let index = index {
+      kbDict = userKeyboards![index] as? [String: String]
+    } else {
+      kbDict = keyboardsDictionary[key]
+    }
+
+    let kbVersion = latestKeyboardFileVersion(withID: keyboardID) ?? "1.0"
+    let isRTL = kbDict?[Key.keyboardRTL]
+
+    var jsFont = jsFontString(font)
+    var jsOskFont: String
+    if oskFont == nil || oskFont == "''" {
+      jsOskFont = jsFont
+    } else {
+      jsOskFont = jsFontString(oskFont)
+    }
+
+    let escapedLangName = languageName!.replacingOccurrences(of: "'", with: "\\'")
+    let escapedKbName = keyboardName!.replacingOccurrences(of: "'", with: "\\'")
+    if jsFont == "''" {
+      jsFont = "undefined"
+    }
+    if jsOskFont == "''" {
+      jsOskFont = "undefined"
+    }
+    let jsString = "setKeymanLanguage('\(escapedKbName)','\(keyboardID)','\(escapedLangName)'," +
+    "'\(languageID)','\(kbVersion)',\(jsFont),\(jsOskFont))"
+    kmLog("Evaluating JavaScript: \(jsString)", checkDebugPrinting: true)
+    inputView.evaluateJavaScript(jsString, completionHandler: nil)
+
+    let userData = isSystemKeyboard ? UserDefaults.standard : activeUserDefaults()
+
+    userData.set([
+      Key.keyboardId: keyboardID,
+      Key.languageId: languageID,
+      Key.keyboardName: keyboardName,
+      Key.languageName: languageName,
+      Key.keyboardVersion: kbVersion,
+      Key.keyboardRTL: isRTL ?? "N",
+      Key.font: jsFont
+      ], forKey: Key.userCurrentKeyboard)
+    userData.synchronize()
+
+    if isKeymanHelpOn {
+      helpBubbleView?.removeFromSuperview()
+      let showHelpBubble = #selector(self.showHelpBubble as () -> Void)
+      NSObject.cancelPreviousPerformRequests(withTarget: self, selector: showHelpBubble, object: nil)
+      perform(showHelpBubble, with: nil, afterDelay: 1.5)
+    }
+
+    let kbInfo = [
+      Key.keyboardId: keyboardID,
+      Key.languageId: languageID,
+      Key.keyboardName: escapedKbName,
+      Key.languageName: escapedLangName,
+      Key.keyboardVersion: kbVersion,
+      Key.font: jsFont
+    ]
+
+    NotificationCenter.default.post(name: .keymanKeyboardChanged, object: self,
+                                    userInfo: [Key.keyboardInfo: kbInfo])
+    return true
+  }
+
+  public func setKeyboard(_ keyboard: [String: String]) -> Bool {
+    return setKeyboard(withID: keyboard[Key.keyboardId]!,
+                       languageID: keyboard[Key.languageId]!,
+                       keyboardName: keyboard[Key.keyboardName]!,
```

Comment:
A new function?  It seems to have been partially extracted from the old `switchToNextKeyboard` function from the Obj-C code.

Reply:
New function to avoid duplicating code that pulls values out of a dictionary. All this will be cleaned up with #388 as well.